client/src/controllers/BulkController.ts

```diff
@@ -82,14 +150,34 @@ export class BulkController extends Controller<HTMLElement> {
   }
 
   /**
-   * Toggles all item checkboxes based on select-all checkbox.
+   * Toggles all item checkboxes, can be used to force check or uncheck all.
+   * If the event used to trigger this method does not have a suitable target,
+   * the first allTarget will be used to determine the current checked value.
    */
-  toggleAll(event: Event & { params?: { force?: boolean } }): void {
-    const force = event?.params?.force;
-    const checkbox = event.target as HTMLInputElement;
-    const isChecked = typeof force === 'boolean' ? force : checkbox.checked;
+  toggleAll(
+    event: CustomEvent<ToggleAllOptions> & { params?: ToggleAllOptions },
+  ) {
+    const { force = null } = {
+      ...event?.detail,
+      ...event?.params,
+    };
```

Comment:
There are a few things I’m confused about here:

- The types say `event` is a `CustomEvent`, while the code below has an optional chaining as if this could be called without an event. But the code later accesses `event.target` anyway.
- Unclear why we need to pass the data via `detail` in addition to `params`. I don’t see either of those two options used in Wagtail aside from the test suite. If we did use them, the use case for `detail` feels very small.
- `force` seems like it doesn’t need a default value (but perhaps worth setting one for clarity?)

Reply:
Thanks @thibaudcolas  - yes, I can see that this is a bit confusing. I would actually love some suggestions how to do this pattern better but let me explain the goals.

The purpose of this odd syntax is so that `toggleAll` can be called a few ways.

1. Internally from the class if needed `this.toggleAll()`
2. Via a simple Stimulus action `<input ... data-action"change->w-bulk#toggleAll" />`
3. Via dispatching some other event to this, also via a Stimulus action `<input ... data-action"some-event@document->w-bulk#toggleAll" />`

However, if we want to use `force`, we can now do this via the actions either with our custom event or with Stimulus action params.

`<input ... data-action"change->w-bulk#toggleAll" data-w-bulk-force-param="true" />`

OR

```
<input ... data-action"some-event@document->w-bulk#toggleAll" />
// JS
document.dispatchEvent(new CustomEvent({ detail: {force:true} }));
```

I could not find a nice way in TypeScript to reflect these three ways of calling, along with have a shared 'options' Type (to contain that `force` could be X type). I would love any tips for this as this approach is super useful in many other controllers.

As for the `force` defaulting to `null`, I do not recall why, it's probably just a habit of mine to always think of the default as an intentional value. Instead of accidental undefined we are better off saying that this may not be passed in and if so it should default to something.